### ISSUE TITLE
Setting the configuration Anchor

### ISSUE THREAD

**Issue by wenjun90** [mnslarcher/kmeans-anchors-ratios#1]
Hi @mnslarcher 
Thank your for your great job.
I have question for setting the ratio of anchor in my dataset that is the detection of text blocs like that:
https://miro.medium.com/max/1200/1*gAx3-sIpo09bPDCZ2fI_kw.png

`anchor_ratios = get_optimal_anchor_ratios(
    ANNOTATIONS_PATH,
    input_size=512,
    scale_bboxes=True,
    num_runs=10,
    num_anchor_ratios=9,
    max_iter=300,
    min_size=0,
    decimals=1,
)`

I chose num_anchor_ration = 9, it's good?

Thank you for your answers.

**Comment by mnslarcher** [comment]
Hi @wenjun90, most of the object detection papers use just 3 ratios (that doesn't mean 3 anchors given that the #anchors = #ratios * #sizes).

From you image and the description of your problem I see that you have very different aspect ratios. Before jumping to 9 I would try to gradually increase the number starting from 3.

I am going to release some code to automatically calculate the % of bounding boxes in a dataset that have an IoU greater than an user-specified threshold, in this way understanding the benefit of increasing the number of ratios will be fast and straightforward. I will need some time because I only work on this during the weekends.

Obliviously consider that more ratios means more parameters in your model/training time/RAM etc., so is better to not exceed.

**Comment by wenjun90** [comment]
@mnslarcher 
Thank you for your answer. 
According to you, I have to change the scale ou keep the default scale: anchors_scales: '[2 ** 0, 2 ** (1.0 / 3.0), 2 ** (2.0 / 3.0)]'.

I train with my problem like that:
`python train.py -c 0 -p dataset --head_only True --lr 1e-3 --batch_size 8 --load_weights weights/efficientdet-d0.pth  --num_epochs 30`

I obtain my result affter 30 epoches:

Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=10.49s).
Accumulating evaluation results...
DONE (t=0.71s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.018
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.054
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.007
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.018
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.023
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.146
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.266
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.266


This is normal ou I need to increase the epoch number?

Thank you

**Comment by mnslarcher** [comment]
> @mnslarcher
> Thank you for your answer.
> According to you, I have to change the scale ou keep the default scale: anchors_scales: '[2 ** 0, 2 ** (1.0 / 3.0), 2 ** (2.0 / 3.0)]'.
> 
> I train with my problem like that:
> `python train.py -c 0 -p dataset --head_only True --lr 1e-3 --batch_size 8 --load_weights weights/efficientdet-d0.pth --num_epochs 30`
> 
> I obtain my result affter 30 epoches:
> 
> Running per image evaluation...
> Evaluate annotation type _bbox_
> DONE (t=10.49s).
> Accumulating evaluation results...
> DONE (t=0.71s).
> Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.018
> Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.054
> Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.007
> Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
> Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
> Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.018
> Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.023
> Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.146
> Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.266
> Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.266
> 
> This is normal ou I need to increase the epoch number?
> 
> Thank you

No, the results are not "normal", your AP are still very low. Is impossible to tell what is the problem just from this output, if you have a plot with training and validation losses (maybe taken from TensorBoard) I can at least help you understand if you are over/under fitting your data.

How large are your train/val datasets?

**Comment by wenjun90** [comment]
[
![image](https://user-images.githubusercontent.com/64893076/82301908-2eb78500-99b9-11ea-9d52-1b1b655708ec.png)
](url)

J'utilise 2500 image train et 300 valid. There is no overfitting

**Comment by mnslarcher** [comment]
Yes, you are not over-fitting. I would try two things:

1) see if you are able to over-fit you dataset, this sound stupid but is a standard practice to double check that you don't have any major problem, if you are able to over-fit you data you can try to increase the regularization and see what happen. If you are not able to decrease the train loss more than this you have a major problem (and maybe is related also to anchor ratios or maybe not)
2) I would also try to run some error analysis, to see if there are specific elements that the model struggle to detect

Basically I would try to understand if there are optimization problems or your model is not complex enough (for example because tuning just the head is not enough), I would say that the first case is more likely
